Add a pre-requisite for XPU windwos build

guangyey opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126752

# Motivation
As mentioned in [[RFC] Enable PyTorch XPU on Native Windows on Intel GPUs](https://github.com/pytorch/pytorch/issues/126719), we propose to enable PyTorch XPU on Native Windows on Intel GPUs. This PR is a pre-requisite intending to support XPU runtime on Windows.
The difference from Linux is that default context feature is disabled by default on Windows. We use a constructed context associated with all devices we enumerated as a workaround. We should switch to the default context when this feature is enabled by default on Windows. Otherwise, the impact is that our context cannot be shared with other components or libraries, which will lead to some features, such as zero-copy, not working on Windows.

gujinghui commented on c10/xpu/XPUFunctions.cpp:
```diff
@@ -53,10 +53,22 @@ inline void initGlobalDevicePoolState() {
+  for (auto it = gDevicePool.devices.begin(); it != gDevicePool.devices.end();
+       ++it) {
+    deviceList.push_back(*(*it));
+  }
```
Is std::transform or std::reduce able to cover these statements?

guangyey replied:
`std::transform` has a matched scenario. But the element stored in `gDevicePool.devices` is a `std::unique_ptr` which does not have a `move constructor` that is a MUST to use within `std::transform` efficiently.

guangyey committed:
Add a pre-requsite for XPU windwos build

[ghstack-poisoned]

guangyey committed:
Update on "Add a pre-requisite for XPU windwos build"


# Motivation
As mentioned in [[RFC] Enable PyTorch XPU on Native Windows on Intel GPUs](https://github.com/pytorch/pytorch/issues/126719), we propose to enable PyTorch XPU on Native Windows on Intel GPUs. This PR is a pre-requisite intending to support XPU runtime on Windows.
The difference from Linux is that default context feature is disabled by default on Windows. We use a constructed context associated with all devices we enumerated as a workaround. We should switch to the default context when this feature is enabled by default on Windows. Otherwise, the impact is that our context cannot be shared with other components or libraries, which will lead to some features, such as zero-copy, not working on Windows.


[ghstack-poisoned]

guangyey committed:
Update on "Add a pre-requisite for XPU windwos build"


# Motivation
As mentioned in [[RFC] Enable PyTorch XPU on Native Windows on Intel GPUs](https://github.com/pytorch/pytorch/issues/126719), we propose to enable PyTorch XPU on Native Windows on Intel GPUs. This PR is a pre-requisite intending to support XPU runtime on Windows.
The difference from Linux is that default context feature is disabled by default on Windows. We use a constructed context associated with all devices we enumerated as a workaround. We should switch to the default context when this feature is enabled by default on Windows. Otherwise, the impact is that our context cannot be shared with other components or libraries, which will lead to some features, such as zero-copy, not working on Windows.


[ghstack-poisoned]

guangyey commented:
merge this PR to https://github.com/pytorch/pytorch/pull/126725

guangyey committed:
Update on "Add a pre-requisite for XPU windwos build"


# Motivation
As mentioned in [[RFC] Enable PyTorch XPU on Native Windows on Intel GPUs](https://github.com/pytorch/pytorch/issues/126719), we propose to enable PyTorch XPU on Native Windows on Intel GPUs. This PR is a pre-requisite intending to support XPU runtime on Windows.
The difference from Linux is that default context feature is disabled by default on Windows. We use a constructed context associated with all devices we enumerated as a workaround. We should switch to the default context when this feature is enabled by default on Windows. Otherwise, the impact is that our context cannot be shared with other components or libraries, which will lead to some features, such as zero-copy, not working on Windows.


[ghstack-poisoned]
